processing/src/main/java/io/druid/segment/filter/JavaScriptFilter.java

```diff
@@ -51,34 +56,59 @@ public ImmutableBitmap getBitmapIndex(final BitmapIndexSelector selector)
   {
     final Context cx = Context.enter();
     try {
-      final Indexed<String> dimValues = selector.getDimensionValues(dimension);
       ImmutableBitmap bitmap;
-      if (dimValues == null) {
-        bitmap = selector.getBitmapFactory().makeEmptyImmutableBitmap();
-      } else {
-        bitmap = selector.getBitmapFactory().union(
-            FunctionalIterable.create(dimValues)
-                              .filter(
-                                  new Predicate<String>()
-                                  {
-                                    @Override
-                                    public boolean apply(@Nullable String input)
-                                    {
-                                      return predicate.applyInContext(cx, input);
-                                    }
-                                  }
-                              )
-                              .transform(
-                                  new com.google.common.base.Function<String, ImmutableBitmap>()
-                                  {
-                                    @Override
-                                    public ImmutableBitmap apply(@Nullable String input)
-                                    {
-                                      return selector.getBitmapIndex(dimension, input);
-                                    }
-                                  }
-                              )
-        );
+
+      boolean allEmptyDimensions = true;
+      Indexed<String>[] dimValuesList = new Indexed[dimensions.length];
+      Iterator<String>[] dimValuesIterator = new Iterator[dimensions.length];
+      String[] currentDim = new String[dimensions.length];
+      for (int idx = 0; idx < dimensions.length; idx++) {
+        dimValuesList[idx] = selector.getDimensionValues(dimensions[idx]);
+        if (dimValuesList[idx] != null && dimValuesList[idx].size() > 0) {
+          allEmptyDimensions = false;
+        } else {
+          dimValuesList[idx] = EMPTY_STR_DIM_VAL;
+        }
+        dimValuesIterator[idx] = dimValuesList[idx].iterator();
+        if (idx != 0) {
+          currentDim[idx] = dimValuesIterator[idx].next();
```

Comment:
`currentDim` is bit misleading, since it represent the current dimension value, and not the current dimension name. `currentDimValue` would be clearer.